thirdparty/faiss/faiss/Clustering.cpp

```diff
@@ -573,8 +557,12 @@ void Clustering::train_encoded(
             double t0s = getmillisecs();
 
             if (!codec) {
-                index.assign(nx, reinterpret_cast<const float *>(x),
-                             assign.get(), dis.get());
+                index.search(
```

Comment:
Early stop happens in (diff = (prev_obj - stats.obj)/ prev_obj, diff < 0.01 * early_stop_threshold).
Update centroids in each iteration can make L2 loss value (obj) become smaller, but not work for IP metric. When metic = IP, early stop in iter = 1 can easily happen in prev.obj < obj, prev.obj - obj < 0.

assign () use L2 distance to avoid this.
Maybe diff = abs(prev_obj - stats.obj)/ prev_obj is better?Early stop happens in (diff = (prev_obj - stats.obj)/ prev_obj, diff < 0.01 * early_stop_threshold).
Update centroids in each iteration can make L2 loss value (obj) become smaller, but not work for IP metric. When metic = IP, early stop in iter = 1 can easily happen in prev.obj < obj, prev.obj - obj < 0.

assign () use L2 distance to avoid this.
Maybe diff = abs(prev_obj - stats.obj)/ prev_obj is better?

Reply:
A good point. Let me see...

Reply:
force-pushed the update. Enforced L2 distance for IP / COSINE in order to support early stop criteria.

Reply:
Did we try how it perform with metric = IP and do Kmeans using L2 vs IP?

Reply:
yes, if metric=IP(cosine) and kmeans=L2, then it provides a higher recall than if kmeans=IP(cosine)